Update systemd unit for Tahoe-LAFS

hacklschorsch opened the pull request:
The Tahoe-LAFS service definition has issues:

  - It tries to create users without a group (a user cannot be created without a group so this fails at evaluate time) and doesn't use them to run tahoe, thus tahoe (would) run as root (if it wasn't too broken to actually start).  This PR uses one user per defined tahoe service. **POSSIBLY BREAKING**: You might have to `chown` your existing files if you somehow managed to run tahoe with the previous configuration.
  - It isn't compatible with the past few Tahoe-LAFS releases (the pidfile config does not work, non-pidfile running requires an extra parameter for tahoe not to exit directly after starting (cf. tahoe-lafs tickets [3921](https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3921), [4036](https://tahoe-lafs.org/trac/tahoe-lafs/ticket/4036), [4159](https://tahoe-lafs.org/trac/tahoe-lafs/ticket/4159)))
  - It did make assumptions about the location of data and configuration in the file system hierarchy that (IIANM) aren't up to date with where NixOS usually puts things (under `/var/db` instead of using [systemd's `StateDirectory` facility](https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html) - **POSSIBLY BREAKING**: If you ran tahoe with the earlier (defunct for quite a while now) service definition, this changing of locations means you have to rename/move your config and data.
  - It makes tahoe listen only on IPv4, not also on IPv6.

This PR integrates changes (hopefully all improvements?) that we already run in "production" (if you want to call it that, test grid is for testing things out) in the  Tahoe-LAFS public test grid ([information about that grid](https://tahoe-lafs.org/trac/tahoe-lafs/wiki/TestGrid#HowToConnectToThePublicTestGrid), [(all-Nix) configuration of that grid](https://github.com/tahoe-lafs/infrastructure/tree/main/testgrid.tahoe-lafs.org)) and fixes the tahoe service definition so one could actually run a tahoe service with it ([like we do](https://github.com/tahoe-lafs/infrastructure/blob/main/testgrid.tahoe-lafs.org/tahoe-lafs-testgrid.nix)).
 
## Things done

  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

hacklschorsch committed:
tahoe: Listen on IPv6 in addition to IPv4

By using a tcp6 socket instead of a tcp4 socket, we
can have a "dual stack" configuration without having
to create two server endpoints in Twisted python.

Ref https://en.wikipedia.org/wiki/IPv6#IPv4-mapped_IPv6_addresses

hacklschorsch committed:
tahoe: Run under own users & closer to systemd recommendations

Create system users for tahoe servers and run
the servers as these users.

Run tahoe in the foreground and without a pidfile.

hacklschorsch committed:
tahoe: Align system users' homes with service data directories

hacklschorsch commented:
Pinging @MostAwesomeDude since you're listed as the maintainer of tahoe-lafs in nixpkgs. 👋
Thank you for your work!

MostAwesomeDude reviewed (approved):
Thanks for preserving and improving the comments.

hacklschorsch committed:
Merge remote-tracking branch 'nixos/master' into update-systemd-unit-for-tahoe-lafs

Update with current master hoping that fixes the unrelated `fmt-check`
error in `lib/.version` ("No final newline expected").

hacklschorsch committed:
Merge remote-tracking branch 'nixos/master' into update-systemd-unit-for-tahoe-lafs

hacklschorsch committed:
Tahoe-LAFS service definition: Add comment about tcp6 socket

The tcp6 socket handles ipv4 connections as well, so this
is a dual-stack configuration instead of the earlier ipv4-only
config.

hacklschorsch commented:
I think this is done - or am I missing something?

hacklschorsch committed:
Merge remote-tracking branch 'nixos/master' into update-systemd-unit-for-tahoe-lafs
